### PR TITLE
Adjust tests with terminal logger enabled

### LINKED ISSUE #9796: [Bug]: Several MSBuild unit tests are failing with terminal logger enabled
### Issue Description

I feel like this is already tracked somewhere but can't find it.

Microsoft.Build.Engine.UnitTests:
```
0.7754982s✘ Microsoft.Build.UnitTests.TerminalLoggerConfiguration_Tests.TerminalLoggerOnByEnv(envVarSource: \"MSBUILDLIVELOGGER\")
Shouldly.ShouldAssertException : superset.Contains(key)\r\n    should be\r\nTrue\r\n    but was\r\nFalse\r\n\r\nAdditional Info:\r\n    environment variable removed: MSBUILDLIVELOGGER
   at Microsoft.Build.UnitTests.EnvironmentInvariant.<AssertInvariant>g__AssertDictionaryInclusion|2_0(IDictionary superset, IDictionary subset, String operation) in C:\src\msbuild\src\UnitTests.Shared\TestEnvironment.cs:line 434
   at Microsoft.Build.UnitTests.EnvironmentInvariant.AssertInvariant(ITestOutputHelper output) in C:\src\msbuild\src\UnitTests.Shared\TestEnvironment.cs:line 424
   at Microsoft.Build.UnitTests.TestEnvironment.Cleanup() in C:\src\msbuild\src\UnitTests.Shared\TestEnvironment.cs:line 95
   at Microsoft.Build.UnitTests.TestEnvironment.Dispose() in C:\src\msbuild\src\UnitTests.Shared\TestEnvironment.cs:line 68
   at Microsoft.Build.UnitTests.TerminalLoggerConfiguration_Tests.Dispose() in C:\src\msbuild\src\Build.UnitTests\TerminalLoggerConfiguration_Tests.cs:line 51
   at ReflectionAbstractionExtensions.DisposeTestClass(ITest test, Object testClass, IMessageBus messageBus, ExecutionTimer timer, CancellationTokenSource cancellationTokenSource) in /_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs:line 79
```

Microsoft.Build.CommandLine.UnitTests:
```
0.1690553s✘ Microsoft.Build.Engine.UnitTests.MSBuildServer_Tests.BuildsWhileBuildIsRunningOnServer
System.AggregateException : One or more errors occurred.\r\n---- System.FormatException : Input string was not in a correct format.\r\n---- A file expected as an output does not exist: C:\\Users\\laprosek\\AppData\\Local\\Temp\\5czuhebl.mqs\\MSBuildTemp\\tmp4057706171da4d4a91df81e654793a2e.tmp\r\nExpected: True\r\nActual:   False
----- Inner Stack Trace #1 (System.FormatException) -----
   at System.Number.StringToNumber(String str, NumberStyles options, NumberBuffer& number, NumberFormatInfo info, Boolean parseDecimal)
   at System.Number.ParseInt32(String s, NumberStyles style, NumberFormatInfo info)
   at System.Int32.Parse(String s)
   at Microsoft.Build.Engine.UnitTests.MSBuildServer_Tests.ParseNumber(String searchString, String toFind) in C:\src\msbuild\src\MSBuild.UnitTests\MSBuildServer_Tests.cs:line 348
   at Microsoft.Build.Engine.UnitTests.MSBuildServer_Tests.BuildsWhileBuildIsRunningOnServer() in C:\src\msbuild\src\MSBuild.UnitTests\MSBuildServer_Tests.cs:line 190
----- Inner Stack Trace #2 (Xunit.Sdk.TrueException) -----
   at Microsoft.Build.UnitTests.TransientTestFile.Revert() in C:\src\msbuild\src\UnitTests.Shared\TestEnvironment.cs:line 691
   at Microsoft.Build.UnitTests.TestEnvironment.Cleanup() in C:\src\msbuild\src\UnitTests.Shared\TestEnvironment.cs:line 89
   at Microsoft.Build.UnitTests.TestEnvironment.Dispose() in C:\src\msbuild\src\UnitTests.Shared\TestEnvironment.cs:line 68
   at Microsoft.Build.Engine.UnitTests.MSBuildServer_Tests.Dispose() in C:\src\msbuild\src\MSBuild.UnitTests\MSBuildServer_Tests.cs:line 90
   at ReflectionAbstractionExtensions.DisposeTestClass(ITest test, Object testClass, IMessageBus messageBus, ExecutionTimer timer, CancellationTokenSource cancellationTokenSource) in /_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs:line 79
Output:
Executing [C:\src\msbuild\artifacts\bin\Microsoft.Build.CommandLine.UnitTests\Debug\net472\MSBuild.exe C:\Users\laprosek\AppData\Local\Temp\5czuhebl.mqs\MSBuildTemp\Temporary6095b3e82d274a3f9165c921da46cf74\testProject.proj]
==== OUTPUT ====
MSBuild version 17.10.0-dev-24128-01+eec991b08 for .NET Framework
\x1b]9;4;3;\x1b\\x1b[?25l\x1b[1F
\x1b[?25h
Build \x1b[32;1msucceeded\x1b[m in 0.0s
\x1b]9;4;0;\x1b\
```

.. and more.

### Steps to Reproduce

```
C:\src\msbuild>SET MSBUILDLIVELOGGER=auto
C:\src\msbuild>build -test
```

### Expected Behavior

Tests succeed.

### Actual Behavior

Some tests fail.

### Analysis

_No response_

### Versions & Configurations

MSBuild at commit 6f44380e4fdea6ddf5c11f48efeb25c2bf181e62

### PR DESCRIPTION
**Pull request by GangWang01**
Fixes #9796

### Context
With terminal logger enabled by `SET MSBUILDLIVELOGGER=auto`, current tests has the following problems.

- `Message` is ignored by terminal logger currently and this causes tests depending on the message fail. Though #9667 could get messages ptinted in terminal logger, it is a way to opt. To make messages output in console logger and terminal logger both, we need the workaround for printing messages in terminal logger. Affected tests: Microsoft.Build.Engine.UnitTests.MSBuildServer_Tests.
- The environment variable `MSBUILDLIVELOGGER`/`MSBUILDTERMINALLOGGER` affects tests for these two environment variables. Affected tests: Microsoft.Build.UnitTests.TerminalLoggerConfiguration_Tests.TerminalLoggerOnByEnv.
- Enabling terminal logger by the environment variable gets extra more build events than other loggers by https://github.com/dotnet/msbuild/blob/bb7846e599da22193f35649bc3cd90073cbf8c8c/src/MSBuild/XMake.cs#L2854-L2855. Affected test: Microsoft.Build.UnitTests.TerminalLogger_Tests.TestTerminalLoggerTogetherWithOtherLoggers.

In addition, cleaning up created TestEnvironment should revert variants in reverse order. 

### Changes Made

- For `Message`, use the workaround adding --interactive for terminal logger to make messages output in console logger and terminal logger both.
- Clear pre-setting environment variables of terminal logger to affect the tests.
- Reverting variants in reverse order while cleaning up created TestEnvironment.

### Testing
Run existing tests locally with setting the environment variable. Also verified in CI through #9957.

### CODE REVIEW

File: src/UnitTests.Shared/TestEnvironment.cs [hunk 1]
```diff
@@ -83,6 +87,8 @@ private void Cleanup()
             {
                 _disposed = true;
 
+                // It should be in reverse order revert to get back to original state.
+                _variants.Reverse();
                 // Reset test variants
                 foreach (var variant in _variants)
                 {
```
Comment (ladipro) [inline src/UnitTests.Shared/TestEnvironment.cs:94]: Curious, is this a pro-active fix or did you actually run into variants that depend on the order?

also, nit:

```suggestion
                // Reset test variants in reverse order to get back to original state.
                for (int i = _variants.Count - 1; i >= 0; i--)
                {
                    var variant = _variants[i];
```

File: src/UnitTests.Shared/TestEnvironment.cs [hunk 2]
```diff
@@ -53,6 +53,10 @@ public static TestEnvironment Create(ITestOutputHelper output = null, bool ignor
                 env.WithInvariant(new BuildFailureLogInvariant());
             }
 
+            // Clear these two environment variables first in case pre-setting affects the test.
+            env.SetEnvironmentVariable("MSBUILDLIVELOGGER", null);
+            env.SetEnvironmentVariable("MSBUILDTERMINALLOGGER", null);
```
Comment (ladipro) [inline src/UnitTests.Shared/TestEnvironment.cs:58]: nit: If we do this, it should be possible to delete the same statements from
https://github.com/dotnet/msbuild/blob/6fec6072f0591bede57a71474ddc08b95e3e4c28/src/Build.UnitTests/TerminalLoggerConfiguration_Tests.cs#L31-L33

File: src/UnitTests.Shared/TestEnvironment.cs (hunk 1 above)
Comment (GangWang01) [reply src/UnitTests.Shared/TestEnvironment.cs:94]: While fixing the issue I run into the problem from the order resetting variants.
Given that with `SET MSBUILDLIVELOGGER=auto` we run tests https://github.com/dotnet/msbuild/blob/6fec6072f0591bede57a71474ddc08b95e3e4c28/src/Build.UnitTests/TerminalLoggerConfiguration_Tests.cs#L139-L142, we need to first clear pre-setting `MSBUILDLIVELOGGER`, then set the value for the test. For the variants keeping the value for environment variable `MSBUILDLIVELOGGER`, they are added by the following order.
||original value|current value|
|-|-|-|
|clear pre-setting|auto|null|
|set the value for the test|null|true|

It should revert the variants in reverse order to get back to original value `auto`.

Comment (ladipro) [reply src/UnitTests.Shared/TestEnvironment.cs:94]: Makes perfect sense, thank you!

**Review by ladipro — APPROVED** [review]

**Review by YuliiaKovalova — APPROVED** [review]

**Pull request merged**